### DISCUSSION TITLE
DriveBase.angle() Optimizations

### DISCUSSION THREAD

**Discussion by DylanBPY** [pybricks/support#1808] [Q&A] [answered]
Hello, it's me again!
This time I need some help with optimizing DriveBase.angle() on my EV3. It seems to be drastically inaccurate and I was wondering if there was any way to possible improve the accuracy of it. 
This is how I set up the drive base and retrieve the angle:
```python
axle_track = 72  # mm
wheel_diameter = 43.2  # mm
wheel_circumference = math.pi * wheel_diameter

left_motor = Motor(Port.A)
right_motor = Motor(Port.D)
robot = DriveBase(left_motor, right_motor, wheel_diameter, axle_track)
current_angle = robot.angle() % 360
```

I think I'm doing everything correctly, but knowing me, it's something really obvious that i didn't catch.
Any help is appreciated!

- Dylan

**Comment by DylanBPY** [comment]
Update on this, I've changed up my robot design to test axle track variations, and it's still way off.

**Comment by laurensvalk** [comment, marked as answer]
The robot determines the `angle` based on the rotation of the motors and the parameters that you give it.

With wide or soft wheels, the parameters aren't usually the exact value you'd find with a ruler, but you'll want to improve them experimentally. See the `Measuring and validating the robot dimensions` section [here](https://docs.pybricks.com/en/latest/robotics.html) for a recipe.

**Reply by DylanBPY** [reply]
This makes it a lot better, but the problem arises when I start making the robot do a combination of driving and turning, the angle seems to get messed up there.

**Reply by laurensvalk** [reply]
Now that you know where the values come from, you'll know that the robot's `angle` method is not a magic bullet. It can be quite useful for some purposes, as long as you keep its limitations in mind. It is generally not suited for keeping track of angles across multiple moves as the wheels may slip, and so on.

**Reply by DylanBPY** [reply]
That's fair, how accurate could I get with the actual gyro?

**Reply by laurensvalk** [reply]
Slightly better, but the external EV3 gyro can't be controlled in the same way we can control the internal gyro in the SPIKE Hub, so there's nothing 'builtin' regarding the gyro for EV3.

**Reply by DylanBPY** [reply]
Ok, thank you!
